feat(twenty-four-hour-time): Implement new 24hr time component

mcelestini opened the pull request:
Add a new 24hr time component for use in PVT Abstraction and Adjudication Form. This component includes the following features:
- flexible input formats for time format
- live restriction of user input to numbers and colon characters only
- validation to ensure only good time values can be entered (i.e. no minute value greater than 59)
- a pop-up time picker for incremental time adjustment
- complete keyboard control and aria-live announcements for screen reader users
- similar look and feel to other HQR and CMS components

See gifs below to see this component in action

Flexible input formats
![Kapture 2022-12-30 at 10 13 48](https://user-images.githubusercontent.com/39271238/210085385-e5164972-13be-4c43-92ca-3c56006a8aba.gif)

Live restriction of user's input
![Kapture 2022-12-30 at 10 05 33](https://user-images.githubusercontent.com/39271238/210085140-40831999-75ec-4646-ad15-8bcb22d0c3bd.gif)

Validation that entered time is a good value
![Kapture 2022-12-30 at 10 15 35](https://user-images.githubusercontent.com/39271238/210085549-c54cdbdd-0db2-4399-9801-28d5d7b6e27f.gif)

Pop-up time picker
![Kapture 2022-12-30 at 10 17 31](https://user-images.githubusercontent.com/39271238/210085690-6ea3cda6-6128-4b4c-ab4d-0a0f265cc83a.gif)

Keyboard control
![Kapture 2022-12-30 at 10 18 47](https://user-images.githubusercontent.com/39271238/210085965-f917ba44-8b56-4464-b61f-6f85e3e0b089.gif)
ff

elizabethso reviewed (approved):
This is a cool feature!

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.css:
```diff
@@ -0,0 +1,203 @@
+.time-picker-header {
+  display: flex;
+  justify-content: space-between;
+  font-weight: bold;
```
In my `pillbox` [PR](https://github.com/Bellese/angular-design-system/pull/354/files), I tried to use `ds-u-...` classes from CMS design system where possible. Is this a convention we wish to continue?

`src/app/modules/pillbox/pillbox.component.html` i.e. 
 - `ds-u-padding-left--1`
 - `ds-u-font-size--base`
 - `ds-u-font-weight--bold`

mcelestini replied:
Nice catch, I'll update these

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+  (click)="stopPropagation($event)"
+  (focusout)="focusLost($event)"
+>
+  <div class="time-picker-header">
```
Any classes I could find from CMS design system around
- padding, margins, layout etc
- visual text size, style etc
... I try to keep in the HTML, rather than re-create in CSS

Although maybe we already have enough inconsistencies that there's not much payoff in being that strict?

But, functionally, or at least clarity/readability, I find I like to keep the padding/spacing styles in the HTML. It helps me keep the effect of all these straight and read them alongside the markup, since I'm usually trying to visualize it alongside the pile of `<div>` as I work.

Then any of the `flex` or further styling I put in "semantically" named CSS styles. I see you're also doing this, and it really helps. I just noticed more padding/margin and text style control in your CSS.

Not really a _request_ for change in here, just something I noticed. Since we're constantly discovering and settling on conventions, and I happened to be working on a new component, figured we could chat about whether it's even worth a convention.

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+          (keyup.enter)="stepTime('hours', 'decrement')"
+        >
+          <fa-icon [icon]="faChevronUp"></fa-icon>
+        </div>
```
Any opposition to a newline after closing `</div>` tags? I personally do that when it doesn't seem to affect how it gets rendered, as it helps me track element separation and grouping.

That said, I tend to skip the newlines if I have multiple, single-line elements with both open and close tags inline.

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+        >
+          <fa-icon [icon]="faChevronUp"></fa-icon>
+        </div>
+        <div *ngFor="let minute of previousMinutes">
```
Not sure why, but I also recently found that putting the `{{ X }}` inline, even when it would absolutely fit with room to spare, somehow made it tougher to visually scan for.

Thx for putting it on its own line. Again, I do it this way unless it adversely affects spacing - I once had it add spaces to button labels.

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+    <div class="selected-time-row">
+      <div
+        class="item-selected"
+      >{{timeHours}}</div>
```
See... I'm already a hypocrite ... but here the elements are so simple, and even the class names are short enough, that I'm kinda 🤷 about newlines 😆

mcelestini replied:
I can do these on one line, no problem

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+        <div
+          class="bumper-icon bottom-icon"
+          tabindex="0"
+          aria-label="Activate enter to increment the hour value"
```
`Activate enter...` ... sticking with The Classics? 😅 

Just saying, JAWS/NVIDIA etc will know the `role="button"` and should read appropriately. I'm 99.9999% confident we should be good with just `increment the hour value`

mcelestini replied:
This snuck through, shame!

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+          aria-label="Activate enter to increment the hour value"
+          role="button"
+          (click)="stepTime('hours', 'increment')"
+          (keyup.enter)="stepTime('hours', 'increment')"
```
Ok... but you specifically watch for only `keyup.enter` ... `keyup.space` [should also work](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#keyboard_interactions)

... that said, seems it's recommended to use <button> or `<input type="button">` instead... which would perhaps let `onclick` [do all the lifting](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#required_javascript_features) of `click` and `keyup` and needing to strictly match with the `aria-label` 
```
For native HTML <button> elements, the button's onclick event fires for mouse clicks and when the user presses Space or Enter while the button has focus.
```

mcelestini replied:
This is something that I struggled with for like a day or so: every time I hit space it would scroll the entire page as well, even when preventing default or stopping propagation. I noticed that a few of our other components and various sites around the web also only seem to respond to enter: not really a great rule but an in-the-moment judgment call I made 😬 

Are you okay if I add it to our existing list of future improvements?

hitjim reviewed (commented):
This is as far as I got for today. I'll go over the `.ts` tomorrow.

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.html:
```diff
@@ -0,0 +1,168 @@
+  </div>  
+</ng-template>
+
+<div aria-live="assertive" class="sr-only">
```
I was afraid you'd need to do this 😆 😭

mcelestini replied:
Yeah, it was as we feared 😢

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts:
```diff
@@ -0,0 +1,309 @@
+        formattedInput = `${hours}:${minutes}`;
+      }
+    }
+    // if colon isn't in 2nd or 3rd spot, then we don't allow it anywhere else
```
👍 comments explaining the "why" near the code in question, not just in the block at the top

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts:
```diff
@@ -0,0 +1,309 @@
+      // calculate hour step
+      const currentHour = Number(this.timeHours);
+      let newHour;
+      if (direction === 'increment' && currentHour < 23) {
```
I really prefer a newline above some block of core, or a group of related operations. Or at the very least after variable declarations, or just above some conditional block.

dianadean commented:
This is cool.  I have a few questions though (on the pop-up time picker)....

- The up/down arrows overlap the numbers...intentional?
- There also seems to be a gradient effect...intentional?
- When the picker is enabled, will the user be able to still type the time into the input field?  
- Are the unselected numbers clickable?  If not, do we need to show them (maybe only the selected number and the up/down arrow is visible if you can't interact with the others)?

mcelestini commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.css:
```diff
@@ -0,0 +1,203 @@
+}
+
+.time-picker-label {
+  flex-grow: 1;
```
Yesss

mcelestini commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts:
```diff
@@ -0,0 +1,309 @@
+    e.stopPropagation();
+  }
+
+  addFocusStyling() {
```
I could try using that, to be honest this was just the first way I got this to work and I didn't look back

mcelestini commented:
> This is cool. I have a few questions though (on the pop-up time picker)....
> 
> * The up/down arrows overlap the numbers...intentional?
> * There also seems to be a gradient effect...intentional?
> * When the picker is enabled, will the user be able to still type the time into the input field?
> * Are the unselected numbers clickable?  If not, do we need to show them (maybe only the selected number and the up/down arrow is visible if you can't interact with the others)?

The overlapping numbers w/ arrows and the gradients are intentional, per HCD. While the picker is enabled, you can still type into the input field, but this will automatically close the picker.

The selected numbers _aren't_ clickable yet, but that's part of a future enhancement list we're tracking

Thanks for taking a look

hitjim commented on src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts:
```diff
@@ -0,0 +1,309 @@
+  };
+  removeValidationError = () => {
+    document.getElementById(`time-container-${this.id}`).classList.remove('time-container-error-shake');
+    this.addFocusStyling();
```
This is kinda why lately I'll:
1. Only handle positioning/layout classes I don't want overridden in the HTML
2. Maybe set some default classes in a `setData` function in the `component.ts`
3. In this function, allow complete override of the default classes. I guess I figure if someone's passing in classes to customize, they are hopefully able to figure out and include the ones that'd be default.

mcelestini replied:
I was able to use ngClass for the focus class, but to get the shaking animation to work I still had to getElementById. However, using ngClass for focus removed a lot of these functions, resulting in a lot less duplication. Thanks y'all for your guidance

mcelestini commented:
I believe I've addressed all the feedback with my recent commits. Thank you everyone for reviewing and helping to make this component the best it can be. Aside from what's been mentioned already, there are some CMS utility classes that don't work within the design system that I plan to circle back to post-monorepo when our team revisits this component for another round of enhancements.

dianadean reviewed (approved):
Looks good to me.

hitjim reviewed (approved):
❤️

cshamrick commented:
:tada: This PR is included in version 5.4.0 :tada:

The release is available on [GitHub release](https://github.com/Bellese/angular-design-system/releases/tag/v5.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
